tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -395,22 +395,27 @@ where
     fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
         let span = ctx.span(id).expect("Span not found, this is a bug");
         let mut extensions = span.extensions_mut();
-
-        let mut buf = String::new();
-        if self.fmt_fields.format_fields(&mut buf, attrs).is_ok() {
-            let fmt_fields = FormattedFields {
-                fields: buf,
-                _format_event: PhantomData::<fn(N)>,
-            };
-            extensions.insert(fmt_fields);
+        if let Some(FormattedFields { ref mut fields, .. }) =
+            extensions.get_mut::<FormattedFields<N>>()
+        {
+            let _ = self.fmt_fields.format_fields(fields, attrs);
+        } else {
+            let mut buf = String::new();
+            if self.fmt_fields.format_fields(&mut buf, attrs).is_ok() {
+                let fmt_fields = FormattedFields {
+                    fields: buf,
+                    _format_event: PhantomData::<fn(N)>,
+                };
+                extensions.insert(fmt_fields);
+            }
         }
     }
 
     fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
         let span = ctx.span(id).expect("Span not found, this is a bug");
         let mut extensions = span.extensions_mut();
         if let Some(FormattedFields { ref mut fields, .. }) =
-            extensions.get_mut::<FormattedFields<Self>>()
+            extensions.get_mut::<FormattedFields<N>>()
```

Comment:
Would you mind opening a separate PR against master with these changes? I'd prefer them to merge separately.

Thanks!